internal/testutil/invariant.go

```diff
@@ -0,0 +1,188 @@
+package testutil
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	cid "github.com/ipfs/go-cid"
+)
+
+const TimestampFormat = "2006-01-02T15:04:05-0700"
+
+func CheckLogFileInvariants(filename string) error {
+	file, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	logs := make([]map[string]interface{}, 0)
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		var log map[string]interface{}
+		if err := json.Unmarshal(scanner.Bytes(), &log); err != nil {
+			return err
+		}
+		// fmt.Println(scanner.Text())
+
+		t, err := time.Parse(TimestampFormat, log["ts"].(string))
+		if err != nil {
+			return err
+		}
+		log["ts"] = t
+		logs = append(logs, log)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return err
+	}
+
+	return CheckInvariants(logs)
+}
+
+func CheckInvariants(logs []map[string]interface{}) error {
+	if err := checkBlockReceivedOncePerSession(logs); err != nil {
```

Comment:
We may need to relax this one.

Reply:
You're right it doesn't really make sense when there are a lot of nodes sending blocks to each other at once, I've removed it

Reply:
We could allow specific tests to state their invariants (expect N% duplicate blocks, etc.).

Reply:
At the moment I'm just running the invariants against a fuzz test (also included in this PR). Were you imagining we'd run the invariants against other tests as well?

Reply:
Ideally, yes. The current tests check to see if we somehow end up in the right state. I was hoping the invariant tests could ensure we get there the right way.

I wouldn't enable them for _all_ tests, but it would be nice if we could just wrap arbitrary tests in a simple:


```go
func TestFoo(t *testing.T) {
    defer invariants.Record().Check(invariant1, invariant2, invariant3...)
    // do the actual test.
}
```

But we can do that after we have some basic invariant testing support merged.

---

Note: It would also be _really_ cool if we could enable invariant logging at runtime with an environment variable. If we did that, we could validate invariants from go-ipfs integration tests. But again, we can do that later.

Reply:
I see that makes sense. It should be fairly easy to do invariant testing against arbitrary tests, but we would want to make sure the test is designed to respect invariants (eg wait for cancels to be sent before shutting down bitswap).

The invariant tests just use the standard logs that bitswap outputs, so we would just need to get the logs at DEBUG level and run them through `checkLogFileInvariants()`